WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesAdapter.kt

```diff
@@ -1,121 +1,77 @@
-            if (!containsNode(it)) {
-                return false
-            }
-        }
-        return true
+    override fun getItemId(position: Int): Long {
+        return position.toLong()
     }
 
-    private fun isSameSelectedList(selectedCategoryIdList: HashSet<Long>): Boolean {
-        return selectedCategoryIdList == selectedCategoryIds
+    override fun getItemViewType(position: Int): Int {
+        return position
     }
 
-    private fun containsNode(categoryNode: CategoryNode): Boolean {
-        categoryNodeList.forEach {
-            if (it.categoryId == categoryNode.categoryId) {
-                return true
-            }
-        }
-        return false
+    override fun onBindViewHolder(holder: PrepublishingCategoriesViewHolder, position: Int) {
+        holder.onBind(items[position])
     }
 
-    class CategoriesViewHolder(
-        private val onCheckedChangeClickListener: ((Long, Boolean) -> Unit),
-        itemView: View
-    ) : RecyclerView.ViewHolder(itemView) {
-        fun bind(row: CategoryNode, selectedCategoryIds: HashSet<Long>) = with(itemView) {
-            itemView.isClickable = true
+    @MainThread
+    fun update(newItems: List<PrepublishingCategoriesListItemUiState>) {
+        val diffResult = DiffUtil.calculateDiff(
+                PrepublishingCategoriesDiffUtils(
+                        items.toList(),
+                        newItems
+                )
+        )
+        items.clear()
+        items.addAll(newItems)
+        diffResult.dispatchUpdatesTo(this)
+    }
 
-            setOnClickListener {
-                prepublishing_category_check.isChecked = !prepublishing_category_check.isChecked
+    private class PrepublishingCategoriesDiffUtils(
+        val oldItems: List<PrepublishingCategoriesListItemUiState>,
+        val newItems: List<PrepublishingCategoriesListItemUiState>
+    ) : DiffUtil.Callback() {
+        override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
+            val oldItem = oldItems[oldItemPosition]
+            val newItem = newItems[newItemPosition]
+            if (oldItem::class != newItem::class) {
+                return false
             }
-            val verticalPadding: Int = prepublishing_category_text.resources.getDimensionPixelOffset(
-                    R.dimen.margin_large
-            )
-            val horizontalPadding: Int = prepublishing_category_text.resources.getDimensionPixelOffset(
-                    R.dimen.margin_extra_large
-            )
 
-            ViewCompat.setPaddingRelative(
-                    prepublishing_category_text,
-                    horizontalPadding * row.level,
-                    verticalPadding,
-                    horizontalPadding,
-                    verticalPadding
-            )
+            return ((oldItem).categoryNode.categoryId ==
+                    (newItem).categoryNode.categoryId) &&
+                    (oldItem.checked == newItem.checked)
```

Comment:
Could you please elaborate why we want to compare "checked/unchecked" in "areItemsTheSame"? It feels like a comparison suitable for "areContentsTheSame". Wdyt?

Reply:
It does belong in `areContentsTheSame` - changed in a983929